### PR TITLE
lua script management

### PR DESCRIPTION
**Pull request by wpferguson**
Added data/lua/script_manager.lua to provide lua script installation and management.  Modified data/luarc to load script manager on start.

script_manager creates two modules, Script Install/Update and Script Controller.  Script Install/Update installs the scripts, from the specified repository, if they don't exist.  Once the scripts are installed the module provides a way to update the scripts with one click.  After the scripts are installed, Script Controller starts.  A button for each script is displayed.  Hovering over the button displays  the documentation from the script. Clicking the button loads the script and runs it.  A preference is updated to record that the script is active. Once a script is active, the button changes to Deactivate to turn off the script. Currently there is no way to deactivate a running script, therefore we mark it as deactivated and don't load it the next time darktable starts.

### CODE REVIEW

**Comment by LebedevRI** [comment]
That should probably go into https://github.com/darktable-org/lua-scripts

File: data/lua/script_manager.lua [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+  {[dt.gui.views.lighttable] = {"DT_UI_CONTAINER_PANEL_LEFT_BOTTOM", 100}},   -- containers
+  dt.new_widget("box") -- widget
+  {
+    orientation = "vertical",
+    repository,
+    action_button,
+  },
+  nil,-- view_enter
+  nil -- view_leave
+)
```
Comment (LebedevRI) [inline data/lua/script_manager.lua]: missing newline at the end of the file

**Comment by houz** [comment]
No, it would have to go into dt proper as it's a means to tie the Lua repo into dt.
I didn't look at the code or try it, it's just a general remark.

**Comment by boucman** [comment]
couple of remarks (I havn't read the whole thing yet)
* you can download archives from github using curl, thus avoiding needing git. maybe see if git exists and if it doesn't, then download the archive (you can probably download any commit/branch)
* we need a branch policy on the lua-scripts before we do that. In particular so that lua script branches follow DT branches. I tried to do that at some point, but I am not sure how "clean" we are wrt compatibility. 


So, overall, i'm ok with the idea but I don't think lua-scripts is ready for that yet. of course @supertobi has the last word on that particular aspect.

**Comment by supertobi** [comment]
I have a little bit mixed feeling about lua script management.

1. It is better and easier to use for the user then the luarc file.

2. I don't like the UI, but I don't think it's not possible to make a better one with Lua at the moment. This UI will not work very  if we get more scripts. It would be nice to have a tree structure, and a search field... Or perhaps move it into the Preferences with checkboxes. 

3. And I'm not sure if our structure of the scripts is in a fixed state. 

4. What about the yeald.lua, we need it for some of the other scripts. Do we need special handling here?

5. Do we want the Lua script examples in there?

6. And the really bad think, is that I don't have to much time at the moment... So please don't wait to long for comments from me and move this forward.

**Comment by boucman** [comment]
I agree mostly with @supertobi, it's too early to integrate, the idea is good though.

We can think of a better UI, I can add widgets if needed, or look into making stuff more scriptable in preferences...

but anyway, the big question is settling the lua-scripts infrastructure. I think this PR should be closed and reopened once the other side is ready enough.

**Comment by wpferguson** [comment]
I'm not sure if this is bad form, commenting on a closed PR.  If it is, I apologize in advance.  

I agree that the lua-scripts infrastructure needs to progress some before adding this in.  I'll offer @supertobi whatever assistance I can with that.

In the meantime, does anyone object if I throw script-manager out on the darktable-user list as a drop in replacement for the user's luarc file?  That way the users have something better than the luarc file, we get some testing, and maybe someone will have a better idea after using it.

And, last and least, I guess I'm now in the lead for the "Ugliest UI Element of the Year" contest?  :-D

**Comment by boucman** [comment]
I would temporarly add it to lua-scripts, less chance o it being lost than on the ML, and we will just rmove it from there once it's ready for darktable proper

**Comment by supertobi** [comment]
Please make a PR to lua-scripts, we can improve it there. And feel free to advertise it everywhere you want.

**Comment by wpferguson** [comment]
Where should I put this?  Should it go into contrib or do we want to put it
in a different directory (tools, later, hold, in.work)?  In it's current
form it's not "self-aware" so if a user cloned the repository and enabled
all the scripts it might try and include itself and set up an endless
loop.  Lua may be smart enough to see that it's already loaded and not
allow this to happen.

**Comment by houz** [comment]
Just as a random remark: I kind of like how Blender handles scripts. It doesn't fetch them from the web for you, but it recognizes those you put into the directory it scans and offers them to be enabled/disabled in the settings. With a search box.

**Comment by wpferguson** [comment]
That's pretty much how I'm using it now.  I have lua linked to my
development directory and then when I fire up DT script controller scans
what's there and I turn stuff on and off as needed.  I could kludge
together a search box using an entry and a button.  I'll have to think
about how to return results

**Pull request closed without merging**